Library/OcVariableLib/LegacyVariable.c

```diff
@@ -0,0 +1,182 @@
+/** @file
+  Methods wrapping calls to legacy NVRAM protocol from OpenCore.
+
+  Copyright (C) 2022, mikebeaton. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-3-Clause
+**/
+
+#include <Uefi.h>
+#include <Protocol/OcVariableRuntime.h>
+#include <Protocol/OcFirmwareRuntime.h>
+#include <Library/OcDirectResetLib.h>
+#include <Library/OcVariableLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+
+STATIC
+EFI_STATUS
+InternalLocateVariableRuntimeProtocol (
+  OC_VARIABLE_RUNTIME_PROTOCOL  **OcVariableRuntimeProtocol
+  )
+{
+  EFI_STATUS  Status;
+
+  ASSERT (OcVariableRuntimeProtocol != NULL);
+
+  Status = gBS->LocateProtocol (
+                  &gOcVariableRuntimeProtocolGuid,
+                  NULL,
+                  (VOID **)OcVariableRuntimeProtocol
+                  );
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCVAR: Locate emulated NVRAM protocol - %r\n", Status));
+    return Status;
+  }
+
+  if ((*OcVariableRuntimeProtocol)->Revision != OC_VARIABLE_RUNTIME_PROTOCOL_REVISION) {
+    DEBUG ((
+      DEBUG_WARN,
+      "OCVAR: Emulated NVRAM protocol incompatible revision %d != %d\n",
+      (*OcVariableRuntimeProtocol)->Revision,
+      OC_VARIABLE_RUNTIME_PROTOCOL_REVISION
+      ));
+    return EFI_UNSUPPORTED;
+  }
+
+  return Status;
+}
+
+VOID
+OcLoadLegacyNvram (
+  IN OC_STORAGE_CONTEXT  *Storage,
+  IN OC_GLOBAL_CONFIG    *Config
+  )
+{
+  EFI_STATUS                    Status;
+  OC_VARIABLE_RUNTIME_PROTOCOL  *OcVariableRuntimeProtocol;
+  OC_FIRMWARE_RUNTIME_PROTOCOL  *FwRuntime;
+  OC_FWRT_CONFIG                FwrtConfig;
+
+  Status = InternalLocateVariableRuntimeProtocol (&OcVariableRuntimeProtocol);
+  if (EFI_ERROR (Status)) {
+    return;
+  }
+
+  //
+  // It is not really required to support boot var routing with emulated NVRAM (since there are
+  // no firmware NVRAM boot vars used outside of OpenCore to avoid trashing), but having working
+  // support is more convenient when switching back and forth between emulated and non-emulated
+  // NVRAM, i.e. one less thing to have to remember to switch, since with this code everything
+  // works as expected with or without RequestBootVarRouting. (Without it, boot entries do not
+  // restore to the right place when RequestBootVarRouting is enabled.)
+  // OpenRuntime.efi must be loaded early, but after OpenVariableRuntimeDxe.efi, for this to work.
+  //
+  if (Config->Uefi.Quirks.RequestBootVarRouting) {
```

Comment:
Ugh, that's a bit too tight. Can we avoid accessing config data outside of the main lib?